README.md
# [User Guide](https://henryssondaniel.github.io/teacup.github.io/)
**Te**sting Fr**a**mework for **C**omm**u**nication **P**rotocols and Web Services with visualization web back-end in Python

[![Build Status](https://travis-ci.com/HenryssonDaniel/teacup-visualization-web-back-python.svg?branch=master)](https://travis-ci.com/HenryssonDaniel/teacup-visualization-web-back-python)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=HenryssonDaniel_teacup-visualization-web-back-python&metric=coverage)](https://sonarcloud.io/dashboard?id=HenryssonDaniel_teacup-visualization-web-back-python)
[![latest release](https://img.shields.io/badge/release%20notes-1.0.0-yellow.svg)](https://github.com/HenryssonDaniel/teacup-visualization-web-back-python/blob/master/CHANGES.rst)
[![PyPi](https://img.shields.io/pypi/v/teacup-visualization-back.svg)](https://pypi.org/project/teacup-visualization-back/)
## What ##
Visualization web back-end written in Python.
## Why ##
This project is needed so that the web front-end can interact with the server side and other Teacup projects.
## How ##
Follow the steps below:
1. pip install teacup-visualization-back
1. [The flask command](http://flask.pocoo.org/docs/1.0/tutorial/factory/#run-the-application)
1. [Configure](http://flask.pocoo.org/docs/1.0/tutorial/deploy/#configure-the-secret-key)
1. [Run the application](http://flask.pocoo.org/docs/1.0/tutorial/deploy/#run-with-a-production-server)

For developers:
1. [Run the application](http://flask.pocoo.org/docs/1.0/tutorial/factory/#run-the-application)

CHANGES.rst
Version 1.0.0
------------

Released 2019-07-03

-   Add account API
-   Add dashboard API

tests/test_account.py
#!/usr/bin/env
# -*- coding: utf-8 -*-
"""Account tests"""

from flask import Flask, Response
from itsdangerous import URLSafeSerializer, URLSafeTimedSerializer
from unittest import mock
from unittest.mock import MagicMock

import flaskr.account
import unittest

ACCOUNT = {"email": "email", "firstName": "firstName", "id": "id", "lastName": "lastName"}
AUTHORIZED = '/api/account/authorized'
CHANGE_PASSWORD = '/api/account/changePassword'
EMAIL_DATA = '{"email": "email"}'
LOG_IN = '/api/account/logIn'
LOG_IN_DATA = '{"email": "email", "password": "password"}'
LOG_OUT = '/api/account/logOut'
PASSWORD_TOKEN = '{"password": "password", "token": "'
POST = 'requests.post'
RECOVER = '/api/account/recover'
SECRET_KEY = 'SECRET_KEY'
SIGN_UP = '/api/account/signUp'
SMTP = 'smtplib.SMTP'
TEST_EMAIL = 'test@teacup.com'
VERIFY = '/api/account/verify/'

app = Flask(__name__)
app.config[SECRET_KEY] = 'dev'
app.config['SERVICE_VISUALIZATION'] = 'http://localhost:8080/mysql'
app.config['SMTP_FROM'] = 'noreply@teacup.com'
app.config['SMTP_HOST'] = 'localhost'
app.config['SMTP_PORT'] = '1025'
app.register_blueprint(flaskr.account.blueprint)


class MockResponse:
    """Mock response"""
    def __init__(self, json_data, status_code) -> None:
        self.json_data = json_data
        self.status_code = status_code

    def json(self) -> str:
        """Json"""
        return self.json_data


class TestInit(unittest.TestCase):
    def __mocked_requests_post_log_in_error(*args, **_) -> MockResponse:
        arg = args[0]
        if CHANGE_PASSWORD in arg or SIGN_UP in arg:
            status_code = 200
        else:
            status_code = 400

        return MockResponse(None, status_code)

    def test_user_required(self) -> None:
        name = "name"

        view = MagicMock()
        view.__name__ = name

        self.assertEqual(flaskr.account.user_required(view).__name__, name)

    def test_authorized(self) -> None:
        client = app.test_client()
        self.__set_session_id(client)
        self.assertEqual(client.get(AUTHORIZED).status_code, 200)

    def test_authorized_false(self) -> None:
        self.assertEqual(app.test_client().get(AUTHORIZED).status_code, 401)

    @mock.patch(POST, return_value=MockResponse(ACCOUNT, 200))
    def test_change_password(self, _) -> None:
        self.assertEqual(app.test_client().post(CHANGE_PASSWORD,
                                                data=PASSWORD_TOKEN + self.__create_timed_token() + '"}').status_code,
                         200)

    @mock.patch(POST, side_effect=__mocked_requests_post_log_in_error)
    def test_change_password_log_in_error(self, _) -> None:
        self.assertEqual(app.test_client().post(CHANGE_PASSWORD,
                                                data=PASSWORD_TOKEN + self.__create_timed_token() + '"}').status_code,
                         400)

    @mock.patch(POST, return_value=Response(status=500))
    def test_change_password_request_error(self, _) -> None:
        self.assertEqual(app.test_client().post(CHANGE_PASSWORD,
                                                data=PASSWORD_TOKEN + self.__create_timed_token() + '"}').status_code,
                         500)

    def test_change_password_token_error(self) -> None:
        self.assertEqual(app.test_client().post(CHANGE_PASSWORD, data='{"token": "test"}').status_code, 403)

    def test_change_password_false(self) -> None:
        client = app.test_client()
        self.__set_session_id(client)
        self.assertEqual(client.post(CHANGE_PASSWORD).status_code, 403)

    @mock.patch(POST, return_value=MockResponse(ACCOUNT, 200))
    def test_log_in(self, _) -> None:
        self.assertEqual(app.test_client().post(LOG_IN, data=LOG_IN_DATA).status_code, 200)

    @mock.patch(POST, side_effect=__mocked_requests_post_log_in_error)
    def test_log_in_error(self, _) -> None:
        self.assertEqual(app.test_client().post(LOG_IN, data=LOG_IN_DATA).status_code, 400)

    def test_log_in_false(self) -> None:
        client = app.test_client()
        self.__set_session_id(client)
        self.assertEqual(client.post(LOG_IN).status_code, 403)

    def test_log_out(self) -> None:
        client = app.test_client()
        self.__set_session_id(client)
        self.assertEqual(client.post(LOG_OUT).status_code, 200)

    def test_log_out_false(self) -> None:
        self.assertEqual(app.test_client().post(LOG_OUT).status_code, 401)

    @mock.patch(POST, return_value=Response())
    @mock.patch(SMTP, return_value=MagicMock())
    def test_recover(self, _, __) -> None:
        self.assertEqual(app.test_client().post(RECOVER, data=EMAIL_DATA).status_code, 200)

    @mock.patch(POST, return_value=Response(status=500))
    def test_recover_error(self, _) -> None:
        self.assertEqual(app.test_client().post(RECOVER, data=EMAIL_DATA).status_code, 500)

    def test_recover_false(self) -> None:
        client = app.test_client()
        self.__set_session_id(client)
        self.assertEqual(client.post(RECOVER).status_code, 403)

    @mock.patch(POST, return_value=MockResponse(ACCOUNT, 200))
    @mock.patch(SMTP, return_value=MagicMock())
    def test_sign_up(self, _, __) -> None:
        self.assertEqual(app.test_client().post(SIGN_UP, data=LOG_IN_DATA).status_code, 200)

    @mock.patch(POST, side_effect=__mocked_requests_post_log_in_error)
    @mock.patch(SMTP, return_value=MagicMock())
    def test_sign_up_log_in_error(self, _, __) -> None:
        self.assertEqual(app.test_client().post(SIGN_UP, data=LOG_IN_DATA).status_code, 400)

    @mock.patch(POST, return_value=Response(status=500))
    def test_sign_up_error(self, _) -> None:
        self.assertEqual(app.test_client().post(SIGN_UP, data=EMAIL_DATA).status_code, 500)

    def test_sign_up_false(self) -> None:
        client = app.test_client()
        self.__set_session_id(client)
        self.assertEqual(client.post(SIGN_UP).status_code, 403)

    @mock.patch(POST, return_value=Response(status=200))
    def test_verify(self, _) -> None:
        self.assertEqual(app.test_client().get(VERIFY + self.__create_token()).data, b'The account have been verified')

    @mock.patch(POST, return_value=Response(status=500))
    def test_verify_error(self, _) -> None:
        self.assertEqual(app.test_client().get(VERIFY + self.__create_token()).data,
                         b'The account could not be verified, please try again later')

    def test_verify_bad_token(self) -> None:
        self.assertEqual(app.test_client().get('%stoken' % VERIFY).data, b'The token is not valid')

    @staticmethod
    def __create_timed_token() -> str:
        return URLSafeTimedSerializer(app.config[SECRET_KEY]).dumps(TEST_EMAIL)

    @staticmethod
    def __create_token() -> str:
        return URLSafeSerializer(app.config[SECRET_KEY]).dumps(TEST_EMAIL)

    @staticmethod
    def __set_session_id(client) -> None:
        with client.session_transaction() as session:
            session['id'] = 'test'


if __name__ == '__main__':
    unittest.main()


tests/test___init__.py
#!/usr/bin/env
# -*- coding: utf-8 -*-
"""Init tests"""

from unittest.mock import MagicMock

import flaskr.__init__
import flaskr.account
import flaskr.dashboard
import unittest


class TestInit(unittest.TestCase):
    def test_create_app(self) -> None:
        self.__verifyFlask(flaskr.__init__.create_app())

    def test_create_app_with_config(self) -> None:
        self.__verifyFlask(flaskr.__init__.create_app(MagicMock()))

    def __verifyFlask(self, flask) -> None:
        self.assertEqual(flask.config.get('SECRET_KEY'), 'dev')
        self.assertEqual(flask.config.get('SERVICE_REPORT'), 'http://localhost:8080/mysql')
        self.assertEqual(flask.config.get('SERVICE_VISUALIZATION'), 'http://localhost:8080/mysql')
        self.assertEqual(flask.config.get('SESSION_TYPE'), 'filesystem')
        self.assertEqual(flask.config.get('SMTP_FROM'), 'noreply@teacup.com')
        self.assertEqual(flask.config.get('SMTP_HOST'), 'localhost')
        self.assertEqual(flask.config.get('SMTP_PORT'), '1025')

        blueprints = flask.blueprints

        self.assertEqual(len(blueprints), 2)
        self.assertIsNotNone(blueprints.get('account'))
        self.assertIsNotNone(blueprints.get('dashboard'))


if __name__ == '__main__':
    unittest.main()


tests/test_dashboard.py
#!/usr/bin/env
# -*- coding: utf-8 -*-
"""Dashboard tests"""

from flask import Flask, Response
from unittest import mock

import flaskr.dashboard
import unittest

DASHBOARD = '/api/dashboard'
GET = 'requests.get'

app = Flask(__name__)
app.config['SECRET_KEY'] = 'dev'
app.config['SERVICE_REPORT'] = 'http://localhost:8080/mysql'
app.register_blueprint(flaskr.dashboard.blueprint)


class MockResponse:
    """Mock response"""
    def __init__(self, json_data, status_code) -> None:
        self.json_data = json_data
        self.status_code = status_code

    def json(self) -> str:
        """Json"""
        return self.json_data


class TestInit(unittest.TestCase):
    @mock.patch(GET, return_value=MockResponse({"sessions": []}, 200))
    def test_dashboard(self, _) -> None:
        client = app.test_client()
        self.__set_session_id(client)
        self.assertEqual(client.get(DASHBOARD).status_code, 200)

    @mock.patch(GET, return_value=Response(status=500))
    def test_dashboard_error(self, _) -> None:
        client = app.test_client()
        self.__set_session_id(client)
        self.assertEqual(client.get(DASHBOARD).status_code, 500)

    def test_dashboard_unauthorized(self) -> None:
        self.assertEqual(app.test_client().get(DASHBOARD).status_code, 401)

    @staticmethod
    def __set_session_id(client) -> None:
        with client.session_transaction() as session:
            session['firstName'] = 'test'
            session['id'] = 'test'
            session['lastName'] = 'test'


if __name__ == '__main__':
    unittest.main()


api.py
#!/usr/bin/env
# -*- coding: utf-8 -*-
"""REST API"""

from flask import Flask
from flask import json
from flask import jsonify
from flask import request
from flask import Response
from flask import session
from flask_cors import CORS
from flask_session import Session

import requests

app = Flask(__name__)
app.config['SESSION_TYPE'] = 'filesystem'

cors = CORS(app)
Session(app)


@app.route('/api/dashboard', methods=['GET'])
@app.route('/api/v1/dashboard', methods=['GET'])
@app.route('/api/v1.0/dashboard', methods=['GET'])
def dashboard() -> Response:
    """Dashboard"""
    if 'id' in session:
        response = requests.get('http://localhost:8080/mysql/api/dashboard', data='{"id":"' + session["id"] + '"}',
                                headers={'content-type': 'application/json'})

        response = jsonify({'firstName': session["firstName"], "lastName": session["lastName"]})
    else:
        response = Response(status=401)

    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response


@app.route('/api/account/logIn', methods=['POST'])
@app.route('/api/v1/account/logIn', methods=['POST'])
@app.route('/api/v1.0/account/logIn', methods=['POST'])
def log_in() -> Response:
    """Log in"""
    response = log_in_data(json.loads(request.data))

    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response


def log_in_data(data) -> Response:
    """Log in with data"""
    response = requests.post('http://localhost:8080/mysql/api/account/logIn', data=json.dumps(data),
                             headers={'content-type': 'application/json'})

    status = response.status_code
    if status == 200:
        content = response.json()

        session['email'] = content["email"]
        session['firstName'] = content["firstName"]
        session['id'] = content["id"]
        session['lastName'] = content["lastName"]

        response = Response()
    else:
        response = Response(status=status)

    return response


@app.route('/api/account/logOut', methods=['POST'])
@app.route('/api/v1/account/logOut', methods=['POST'])
@app.route('/api/v1.0/account/logOut', methods=['POST'])
def log_out() -> Response:
    """Log out"""
    if 'id' in session:
        session.pop('id')

    response = Response()
    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response


@app.route('/api/account/recover', methods=['POST'])
@app.route('/api/v1/account/recover', methods=['POST'])
@app.route('/api/v1.0/account/recover', methods=['POST'])
def recover() -> Response:
    """Recover account"""
    if requests.post('http://localhost:8080/mysql/api/account/recover', data=json.dumps(json.loads(request.data)),
                     headers={'content-type': 'application/json'}).status_code == 200:
        response = jsonify({'some': 'data'})
    else:
        response = Response(status=401)

    response.headers.add('Access-Control-Allow-Origin', '*')

    return response


@app.route('/api/account/signUp', methods=['POST'])
@app.route('/api/v1/account/signUp', methods=['POST'])
@app.route('/api/v1.0/account/signUp', methods=['POST'])
def sign_up() -> Response:
    """Sign up"""
    data = json.loads(request.data)

    if requests.post('http://localhost:8080/mysql/api/account/signUp', data=json.dumps(data),
                     headers={'content-type': 'application/json'}).status_code == 200:
        response = log_in_data({"email": data["email"], "password": data["password"]})
    else:
        response = Response(status=401)

    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response


if __name__ == '__main__':
    app.run()


flaskr/__init__.py
#!/usr/bin/env
# -*- coding: utf-8 -*-
"""Initialize the application"""

from flask import Flask
from flask_cors import CORS
from flask_session import Session


def create_app(test_config=None) -> Flask:
    """Create the application"""
    app = Flask(__name__, instance_relative_config=True)
    app.config.from_mapping(
        SECRET_KEY='dev',
        SERVICE_REPORT='http://localhost:8080/mysql',
        SERVICE_VISUALIZATION='http://localhost:8080/mysql',
        SESSION_TYPE='filesystem',
        SMTP_FROM='noreply@teacup.com',
        SMTP_HOST='localhost',
        SMTP_PORT='1025'
    )

    if test_config is None:
        app.config.from_pyfile('config.py', silent=True)
    else:
        app.config.from_mapping(test_config)

    CORS(app)
    Session(app)

    from . import account, dashboard
    app.register_blueprint(account.blueprint)
    app.register_blueprint(dashboard.blueprint)

    return app


flaskr/dashboard.py
#!/usr/bin/env
# -*- coding: utf-8 -*-
"""Dashboard API"""

from flask import Blueprint, current_app as app, jsonify, Response, session
from flaskr.account import user_required

import requests

blueprint = Blueprint('dashboard', __name__, url_prefix='/api')


@blueprint.route('/dashboard', methods=['GET'])
@blueprint.route('/v1/dashboard', methods=['GET'])
@blueprint.route('/v1.0/dashboard', methods=['GET'])
@user_required
def dashboard() -> [Response, int]:
    """Dashboard"""
    response = requests.get(app.config['SERVICE_REPORT'] + '/api/session/summary')

    content = {"account": {'firstName': session["firstName"], "lastName": session["lastName"]}}

    status = response.status_code
    if status == 200:
        content["sessions"] = response.json()["sessions"]

    response = jsonify(content)
    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response, status


flaskr/account.py
#!/usr/bin/env
# -*- coding: utf-8 -*-
"""Account API"""

from email.message import EmailMessage
from flask import Blueprint, current_app as app, json, request, Response, session
from itsdangerous import BadSignature
from itsdangerous import URLSafeSerializer, URLSafeTimedSerializer

import requests
import smtplib

blueprint = Blueprint('account', __name__, url_prefix='/api')


def user_required(view) -> Response:
    """User required"""
    def wrapped_user_view(**kwargs) -> Response:
        """Wrapper view"""
        if 'id' in session:
            return view(**kwargs)

        response = Response(status=401)
        response.headers.add('Access-Control-Allow-credentials', 'true')

        return response

    wrapped_user_view.__name__ = view.__name__
    return wrapped_user_view


def __no_user_required(view) -> Response:
    """No user required"""
    def wrapped_no_user_view(**kwargs) -> Response:
        """Wrapper view"""
        if 'id' in session:
            response = Response(status=403)
            response.headers.add('Access-Control-Allow-credentials', 'true')

            return response

        return view(**kwargs)

    wrapped_no_user_view.__name__ = view.__name__
    return wrapped_no_user_view


@blueprint.route('/account/authorized', methods=['GET'])
@blueprint.route('/v1/account/authorized', methods=['GET'])
@blueprint.route('/v1.0/account/authorized', methods=['GET'])
@user_required
def authorized() -> Response:
    """Log in"""
    response = Response()
    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response


@blueprint.route('/account/changePassword', methods=['POST'])
@blueprint.route('/v1/account/changePassword', methods=['POST'])
@blueprint.route('/v1.0/account/changePassword', methods=['POST'])
@__no_user_required
def change_password() -> Response:
    """Log in"""
    data = json.loads(request.data)

    try:
        email = URLSafeTimedSerializer(app.config['SECRET_KEY']).loads(data["token"], max_age=3600)
        password = data["password"]

        status = requests.post(app.config['SERVICE_VISUALIZATION'] + '/api/account/changePassword',
                               data=json.dumps({"authorized": False, "email": email, "password": password}),
                               headers={'content-type': 'application/json'}).status_code

        if status == 200:
            status = __log_in_data({"email": email, "password": password})
    except BadSignature:
        status = 403

    response = Response(status=status)
    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response


@blueprint.route('/account/logIn', methods=['POST'])
@blueprint.route('/v1/account/logIn', methods=['POST'])
@blueprint.route('/v1.0/account/logIn', methods=['POST'])
@__no_user_required
def log_in() -> Response:
    """Log in"""
    response = Response(status=__log_in_data(json.loads(request.data)))
    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response


@blueprint.route('/account/logOut', methods=['POST'])
@blueprint.route('/v1/account/logOut', methods=['POST'])
@blueprint.route('/v1.0/account/logOut', methods=['POST'])
@user_required
def log_out() -> Response:
    """Log out"""
    session.pop('id')

    response = Response()
    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response


@blueprint.route('/account/recover', methods=['POST'])
@blueprint.route('/v1/account/recover', methods=['POST'])
@blueprint.route('/v1.0/account/recover', methods=['POST'])
@__no_user_required
def recover() -> Response:
    """Recover account"""
    data = json.loads(request.data)

    status = requests.post(app.config['SERVICE_VISUALIZATION'] + '/api/account/recover', data=json.dumps(data),
                           headers={'content-type': 'application/json'}).status_code

    if status == 200:
        email = data["email"]
        __send_email("The recover code: " + URLSafeTimedSerializer(app.config['SECRET_KEY']).dumps(email),
                     'Recover', email)

    response = Response(status=status)
    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response


@blueprint.route('/account/signUp', methods=['POST'])
@blueprint.route('/v1/account/signUp', methods=['POST'])
@blueprint.route('/v1.0/account/signUp', methods=['POST'])
@__no_user_required
def sign_up() -> Response:
    """Sign up"""
    data = json.loads(request.data)

    status = requests.post(app.config['SERVICE_VISUALIZATION'] + '/api/account/signUp', data=json.dumps(data),
                           headers={'content-type': 'application/json'}).status_code
    if status == 200:
        email = data["email"]

        __send_email("Please verify your account by clicking here: " + request.url_root + "api/account/verify/" +
                     URLSafeSerializer(app.config['SECRET_KEY']).dumps(email), 'Verify', email)

        status = __log_in_data({"email": email, "password": data["password"]})

    response = Response(status=status)
    response.headers.add('Access-Control-Allow-credentials', 'true')

    return response


@blueprint.route('/account/verify/<token>', methods=['GET'])
@blueprint.route('/v1/account/verify/<token>', methods=['GET'])
@blueprint.route('/v1.0/account/verify/<token>', methods=['GET'])
def verify(token) -> str:
    """Verify"""
    try:
        email = URLSafeSerializer(app.config['SECRET_KEY']).loads(token)
        status = requests.post(app.config['SERVICE_VISUALIZATION'] + '/api/account/verify',
                               data=json.dumps({"email": email}),
                               headers={'content-type': 'application/json'}).status_code

        if status == 200:
            message = 'The account have been verified'
        else:
            message = 'The account could not be verified, please try again later'
    except BadSignature:
        message = 'The token is not valid'

    return message


def __log_in_data(data) -> int:
    """Log in with data"""
    response = requests.post(app.config['SERVICE_VISUALIZATION'] + '/api/account/logIn', data=json.dumps(data),
                             headers={'content-type': 'application/json'})

    status = response.status_code
    if status == 200:
        content = response.json()

        session['email'] = content["email"]
        session['firstName'] = content["firstName"]
        session['id'] = content["id"]
        session['lastName'] = content["lastName"]

    return status


def __send_email(content, subject, to) -> None:
    """Send email"""
    email_message = EmailMessage()
    email_message.set_content(content)

    email_message['Subject'] = subject + ' your Teacup account'
    email_message['From'] = app.config['SMTP_FROM']
    email_message['To'] = to

    smtp = smtplib.SMTP(app.config['SMTP_HOST'], app.config['SMTP_PORT'])
    smtp.send_message(email_message)
    smtp.quit()
